File: Data Structures & Algorithms/maximum-product-subarray/submission-2.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        # brute force: nested for loop with i and j O(n2)
        # calc product bw i and j: O(n) , total : O(n3)
        mn, mx, ans = nums[0], nums[0], nums[0]
        for num in nums[1:]:
            tmpMx = mx
            mx = max(num, num*tmpMx, num*mn)
            mn = min(num, num*tmpMx, num*mn)
            ans = max(ans, mx)
        return ans

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-2.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int ch[256]={0}, l=0;
        int ans=0;
        for(int r=0; r< s.size(); r++) {
            ch[s[r]]++;
            while(l<r && ch[s[r]]>1) {
                ch[s[l]]--;
                l++;
            } 
            ans=max(ans, r-l+1);
        }
        return ans;
    }
};
